### DISCUSSION TITLE
Pydantic serialization of classes with inheritance

### DISCUSSION THREAD

**Discussion by vaibhav-46** [pydantic/pydantic#10994] [General]
I am trying to setup a node library to be able to run different stuff, and link the outputs between the nodes to create a graph. The basic structure of the nodes, graphs and the individual nodes is as follows : 

**node.py** : 
```
class NodeOutputReference(BaseModel):
    node: "Node"
    output_key: str
    
class Node(BaseModel):
    id: str = Field(default_factory=utils.generate_uuid)
    name: str | None = None
    input_mapping: dict[str, NodeOutputReference] = {}
```
 
 llm.py : 
```
class OpenAI(Node):

    name: str | None = "LLM"
    model: str
    prompt: Prompt | None = None
    group: NodeGroup = NodeGroup.LLMS
```
  
 google_search_tool.py: 
  
```
class GoogleSearchTool(Node):
    """A tool for scraping web pages using the Firecrawl service."""

    group: NodeGroup = NodeGroup.TOOLS
    name: str = "Google Search Tool"
    node_type: Literal["GoogleSearch"] = "GoogleSearch"
```
```
class Agent(Node):

    llm: Node = Field(..., description="LLM used by the agent.")
    group: NodeGroup = NodeGroup.AGENTS

    tools: list[Node] = []
    name: str = "AI Agent"
```

In addition to the above, there are 20+ Node definitions but keeping it down to these two to explain the problem. I would like to do the following, such that the resultant graph created has the right node types. The issue is that in most cases, the Node class object is getting instantiated and not the actual node type.

```

search_node = GoogleSearchTool(input="What is the tallest mountain in the world")
result_summarizer = OpenAI(model="gpt-4o", prompt=prompt, input_mapping={"search_result": NodeOutputReference(search_node, "search_results"))
graph.add_nodes([search_node, result_summarizer])
graph_dict = graph.model_dump(serialize_as_any=True)
new_graph = graph.model_validate(graph_dict)
```

Effort 1 : 
I tried adding a node_type literal to each node class, and specified discriminator=True in the graph.py file. To be able to get this solution to work, I had to add a NodeTypeUnion class which would contain a reference to all the possible Node types. This solution partially worked, but it failed when trying to serialize the NodeOutputReference seen within the Node class. The dependency here became a little unmanageable.

Agent / GoogleSearchTool / LLM  ----> depends on ------> Node, NodeOutputReference, NodeTypeUnion
Node ------> depends on ---------> NodeOutputReference
NodeOutputReference ------> depends on -------> NodeTypeUnion
NodeTypeUnion -----> depends on ------------> Agent / GoogleSearchTool / LLM

I tried using forward references and model_rebuild to solve this, but was unsuccessful in using those to get this to work.

Effort 2 : 

Instead of using a discriminator, I tried to create a node_registry and override the model_validate function to create the appropriate node objects but that lead to a infinite recursion issue. The only alternate solution I could think of here was to serialize each field defined manually. 

Could someone help give me some pointers on how I can go about solving this issue? Have looked at a lot of discussions in this forum, but haven't found anything that worked.

Pydantic version 
-------------------
pydantic version: 2.8.0
 pydantic-core version: 2.20.0
 pydantic-core build: profile=release pgo=true

**Comment by bruno-f-cruz** [comment]
I am not sure this solution is relevant, but if you consider that all definitions are in the list of nodes, you can potentially solve everything else by references to that list by consider that each node's id is unique:

Otherwise I am also curious if there is a way to solve the recursion problem with these kinds of models.

Hope it helps!

```python 
from pydantic import BaseModel, Field
import random
from typing import Literal, Annotated, Union, Generic, TypeVar, Type
from enum  import Enum


TNode = TypeVar("TNode", bound="Node")

class NodeReference(BaseModel):
    id: int
    # output_key: str not sure what this is for

    #def node_from_id(self, graph) -> "Node":
    #    ...

class Node(BaseModel):
    id: int = Field(default=random.randint(0, 1000000))
    name: str | None = None
    input_mapping: dict[str, NodeReference] = {}
    node_type: str

    def to_reference(self) -> NodeReference:
        return NodeReference(id=self.id)

    def __eq__(self, value):
        # sprint(type(self), type(value))
        return type(self) is type(value) and self.id == value.id

class NodeGroup(str, Enum):
    LLMS = "LLMs"
    TOOLS = "Tools"
    AGENTS = "Agents"

class OpenAI(Node):

    name: str | None = "LLM"
    model: str
    prompt: str | None = None
    group: NodeGroup = NodeGroup.LLMS
    node_type: Literal["OpenAI"] = "OpenAI"

class GoogleSearchTool(Node):
    """A tool for scraping web pages using the Firecrawl service."""

    group: NodeGroup = NodeGroup.TOOLS
    name: str = "Google Search Tool"
    node_type: Literal["GoogleSearch"] = "GoogleSearch"


class Agent(Node):

    llm: NodeReference = Field(..., description="LLM used by the agent.")
    group: NodeGroup = NodeGroup.AGENTS

    tools: list[NodeReference] = []
    name: str = "AI Agent"
    node_type: Literal["Agent"] = "Agent"


NodeGroupType = Annotated[Union[Agent, OpenAI, GoogleSearchTool], Field(discriminator="node_type")]


class Graph(BaseModel):
    nodes: list[NodeGroupType] = []
    edges: list[tuple[NodeReference, NodeReference]] = []

    def add_node(self, node: NodeGroupType):
        self.nodes.append(node)

    def add_edge(self, source: NodeReference | NodeGroupType, target: NodeReference | NodeGroupType):
        if isinstance(source, Node):
            source = source.to_reference()
        if isinstance(target, Node):
            target = target.to_reference()
        self.edges.append((source, target))

    #def solve_references(self):
    #    ...


graph = Graph()
llm = OpenAI(model="GPT-3")
graph.add_node(llm)

search_tool = GoogleSearchTool()
graph.add_node(search_tool)

agent = Agent(llm=llm.to_reference())
graph.add_node(agent)

graph.add_edge(llm, agent)
graph.add_edge(search_tool, agent)

print(graph)

ser = graph.model_dump()
print(ser)
deser = Graph.model_validate(ser)

print(graph == deser)

```
